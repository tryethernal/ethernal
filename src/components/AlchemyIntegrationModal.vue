<template>
<v-dialog v-model="dialog" max-width="700" persistent>
    <v-card>
        <v-card-title class="headline">Manage Alchemy Integration</v-card-title>
        <v-card-text>
            When enabled, this integration will give you a webhook that you should add in your Alchemy dashboard, in the Notify section, under "Mined Transactions Notifications". Once this is done, all transactions going through the selected Alchemy app will be synchronized with Ethernal, letting you used all of the features on your dapp!
            <v-divider></v-divider>
            <div v-if="canUseIntegration()">
                <v-switch id="webhookSwitch" :disabled="loading" :loading="loading" v-model="enabled" :label="enabled ? 'Enabled' : 'Disabled'" @change="toggleSwitch"></v-switch>
                <v-text-field id="webhook" append-icon="mdi-content-copy" readonly @click:append="copyWebhook()" outlined dense hide-details="auto" :value="formattedWebhook" v-show="token"></v-text-field>
                <input type="hidden" id="copyElement" :value="formattedWebhook">
            </div>
            <div v-else class="mt-1">
                <v-alert type="warning">This integration can only be used with a workspace connected to an Alchemy endpoint. Create a new workspace connected to your Alchemy app to use it</v-alert>
            </div>
        </v-card-text>

        <v-card-actions>
            <v-spacer></v-spacer>
            <v-btn color="primary" :disabled="loading" text @click="close()">Close</v-btn>
        </v-card-actions>
    </v-card>
</v-dialog>
</template>
<script>
import { mapGetters } from 'vuex';

const API_ROOT_URL = process.env.VUE_APP_API_ROOT;
export default {
    name: 'AlchemyIntegrationModal',
    data: () => ({
        dialog: false,
        resolve: null,
        reject: null,
        loading: false,
        enabled: false,
        token: null,
        options: {}
    }),
    methods: {
        canUseIntegration: function() {
            const server = new URL(this.currentWorkspace.rpcServer);
            return server.host.indexOf('alchemyapi.io') > -1 || server.host.indexOf('alchemy.com') > -1;
        },
        copyWebhook: function() {
            const webhookField = document.querySelector('#copyElement');
            webhookField.setAttribute('type', 'text');
            webhookField.select();

            try {
                const copied = document.execCommand('copy');
                const message = copied ? 'Webhook copied!' : `Couldn't copy webhook`;
                alert(message);
            } catch(error) {
                alert(`Couldn't copy webhook`);
            } finally {
                webhookField.setAttribute('type', 'hidden');
                window.getSelection().removeAllRanges();
            }
        },
        open: function(options = {}) {
            this.options = options || {};
            this.enabled = !!options.enabled;
            if (this.enabled && !this.token) {
                this.loading = true;
                this.server.getWorkspaceApiToken(this.currentWorkspace.name)
                    .then(({data}) => this.token = data.token)
                    .finally(() => this.loading = false);
            }
            this.dialog = true;
            return new Promise((resolve, reject) => {
                this.resolve = resolve;
                this.reject = reject;
            })
        },
        close: function() {
            this.resolve(!!this.token);
            this.reset();
        },
        reset: function() {
            this.options = {};
            this.loading = false;
            this.dialog = false;
            this.resolve = null;
            this.reject = null;
        },
        toggleSwitch: function(value) {
            this.loading = true;
            if (value) {
                this.server.enableAlchemyWebhook()
                    .then(({data}) => this.token = data.token)
                    .finally(() => this.loading = false);
            }
            else {
                this.server.disableAlchemyWebhook()
                    .then(() => this.token = null)
                    .finally(() => this.loading = false);
            }
        }
    },
    computed: {
        ...mapGetters([
            'currentWorkspace'
        ]),
        formattedWebhook: function() {
            return `${API_ROOT_URL}/webhooks/alchemy?token=${this.token}`;
        }
    }
}
</script>
