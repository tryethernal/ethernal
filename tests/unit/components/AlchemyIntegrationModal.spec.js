import MockHelper from '../MockHelper';

import AlchemyIntegrationModal from '@/components/AlchemyIntegrationModal.vue';

describe('AlchemyIntegrationModal.vue', () => {
    let helper;

    it('Should display a message if you are not in an Alchemy workspace', async () => {
        helper = new MockHelper({ currentWorkspace: { rpcServer: 'http://localhost:8545' }});

        const wrapper = helper.mountFn(AlchemyIntegrationModal);

        wrapper.vm.open();
        await wrapper.vm.$nextTick();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should let you enable a webhook', async () => {
        helper = new MockHelper({ currentWorkspace: { rpcServer: 'http://alchemyapi.io/1234' }});
        const enableAlchemyWebhookMock = jest.spyOn(helper.mocks.server, 'enableAlchemyWebhook');

        const wrapper = helper.mountFn(AlchemyIntegrationModal);

        wrapper.vm.open();
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.token).toBeFalsy();
        expect(wrapper.find('#webhook').isVisible()).toBe(false);

        await wrapper.find('#webhookSwitch').trigger('click');
        await wrapper.vm.$nextTick();

        expect(wrapper.find('#webhook').element.value).toMatch(/token=123456abcdef/);

        expect(enableAlchemyWebhookMock).toHaveBeenCalled();
        expect(wrapper.vm.token).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should let you disable a webhook', async () => {
        helper = new MockHelper({ currentWorkspace: { rpcServer: 'http://alchemyapi.io/1234' }});
        const disableAlchemyWebhookMock = jest.spyOn(helper.mocks.server, 'disableAlchemyWebhook');

        const wrapper = helper.mountFn(AlchemyIntegrationModal);

        wrapper.vm.open({ enabled: true });
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.token).toBeTruthy();
        expect(wrapper.find('#webhook').isVisible()).toBe(true);
        
        await wrapper.find('#webhookSwitch').trigger('click');
        await wrapper.vm.$nextTick();

        expect(disableAlchemyWebhookMock).toHaveBeenCalled();
        expect(wrapper.vm.token).toBeFalsy();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
