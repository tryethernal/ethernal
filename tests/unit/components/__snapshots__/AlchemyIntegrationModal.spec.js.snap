// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlchemyIntegrationModal.vue Should display a message if you are not in an Alchemy workspace 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active v-dialog--persistent" style="transform-origin: center center; max-width: 700px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Manage Alchemy Integration</div>
          <div class="v-card__text">
            When enabled, this integration will give you a webhook that you should add in your Alchemy dashboard, in the Notify section, under "Mined Transactions Notifications". Once this is done, all transactions going through the selected Alchemy app will be synchronized with Ethernal, letting you used all of the features on your dapp!
            <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
            <div class="mt-1">
              <div role="alert" class="v-alert v-sheet theme--dark warning">
                <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark"></i>
                  <div class="v-alert__content">This integration can only be used with a workspace connected to an Alchemy endpoint. Create a new workspace connected to your Alchemy app to use it</div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`AlchemyIntegrationModal.vue Should let you disable a webhook 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active v-dialog--persistent" style="transform-origin: center center; max-width: 700px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Manage Alchemy Integration</div>
          <div class="v-card__text">
            When enabled, this integration will give you a webhook that you should add in your Alchemy dashboard, in the Notify section, under "Mined Transactions Notifications". Once this is done, all transactions going through the selected Alchemy app will be synchronized with Ethernal, letting you used all of the features on your dapp!
            <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
            <div>
              <div class="v-input theme--light v-input--selection-controls v-input--switch">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-input--selection-controls__input"><input aria-checked="false" id="webhookSwitch" role="switch" type="checkbox" aria-disabled="false" value="">
                      <div class="v-input--selection-controls__ripple"></div>
                      <div class="v-input--switch__track theme--light"></div>
                      <div class="v-input--switch__thumb theme--light">
                        <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                      </div>
                    </div><label for="webhookSwitch" class="v-label theme--light" style="left: 0px; position: relative;">Disabled</label>
                  </div>
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
              <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--is-readonly v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined" style="display: none;">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><input id="webhook" readonly="readonly" type="text"></div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link mdi mdi-content-copy theme--light"></button></div>
                    </div>
                  </div>
                </div>
              </div> <input type="hidden" id="copyElement" value="http://localhost:8081/webhooks/alchemy?token=null">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`AlchemyIntegrationModal.vue Should let you enable a webhook 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active v-dialog--persistent" style="transform-origin: center center; max-width: 700px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Manage Alchemy Integration</div>
          <div class="v-card__text">
            When enabled, this integration will give you a webhook that you should add in your Alchemy dashboard, in the Notify section, under "Mined Transactions Notifications". Once this is done, all transactions going through the selected Alchemy app will be synchronized with Ethernal, letting you used all of the features on your dapp!
            <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
            <div>
              <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--switch primary--text">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-input--selection-controls__input"><input aria-checked="true" id="webhookSwitch" role="switch" type="checkbox" aria-disabled="false" value="">
                      <div class="v-input--selection-controls__ripple primary--text"></div>
                      <div class="v-input--switch__track theme--light primary--text"></div>
                      <div class="v-input--switch__thumb theme--light primary--text">
                        <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                      </div>
                    </div><label for="webhookSwitch" class="v-label theme--light" style="left: 0px; position: relative;">Enabled</label>
                  </div>
                  <div class="v-messages theme--light primary--text">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
              <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--is-readonly v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined" style="">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><input id="webhook" readonly="readonly" type="text"></div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link mdi mdi-content-copy theme--light"></button></div>
                    </div>
                  </div>
                </div>
              </div> <input type="hidden" id="copyElement" value="http://localhost:8081/webhooks/alchemy?token=123456abcdef">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
